### PR TITLE
Upgrade to version 3.8: Upload pipeline updates per step level and add addtl. SRST2 output

### PR DESCRIPTION
**Pull request by lvreynoso**
See #174 and #175 

## Upload pipeline updates per step level (from #172 and #174)
For each stage, create a [stage name}_status.json which gets written to when a step starts running or is finished. Will be used to get step-level updates for the pipeline visualization.

Requires a new name field in the dag_json to ensure that each stage's status file is unique and will not overwrite a previous stage's status file. (See chanzuckerberg/idseq-web#2440, which must be merged before this PR is merged)

Changes beyond original pull request:

- Added more error logging
- Change location of initial update
- Write a start_time field

## Add additional SRST2 output (from #175)
- `amr_results.csv` will now include three extra columns per gene: `total_reads`, `rpm`, and `dpm`.
- included an updated example dag for the `srst2` step
- Bumped Pipeline version to 3.8

Tested with `examples/amr_human_paired_sample_with_rpm.json`

### CODE REVIEW

File: idseq_dag/steps/run_srst2.py [hunk 1]
```diff
@@ -32,31 +33,33 @@ def run(self):
                 PipelineStepRunSRST2.fill_file_path(f)
         else:
             # Post processing of amr data
+            self.generate_mapped_reads_table()
```
Comment (MarkAZhang) [inline idseq_dag/steps/run_srst2.py]: At line 32, should that be changed to `[2:6]`? Since you now have an additional output file.

Comment (lvreynoso) [reply idseq_dag/steps/run_srst2.py]: Yes, good catch!

File: idseq_dag/steps/run_srst2.py [hunk 2]
```diff
@@ -69,6 +72,76 @@ def get_common_params(self):
...
+            If a new resistance gene db is used, the .bed file will need to be updated manually."""
+        bed_file_path = fetch_from_s3(self.additional_files["resist_genome_bed"], self.output_dir_local, allow_s3mi=False)
+        bedtools_params = ['bedtools', 'coverage', '-b', self.output_files_local()[5], '-a', bed_file_path, '>>', os.path.join(self.output_dir_local, 'matched_reads.tsv')]
+        command.execute(" ".join(bedtools_params))
+
+    def get_total_reads(self, is_zipped, is_fasta):
+        """Gets the total number of reads in the sample by counting them directly from the
+            fastq or fasta files."""
+        input_filenames = self.input_files_local[0]
+        if is_zipped:
```
Comment (MarkAZhang) [inline idseq_dag/steps/run_srst2.py]: Can you add a TODO to factor this out into a utility function? `nonhost_fastq` has a similar function.

**Comment by lvreynoso** [comment]
I just realized Tiago wanted to keep these in separate PRs. Oops. Let me fix that.

File: idseq_dag/steps/run_srst2.py [hunk 3]
```diff
@@ -69,6 +72,76 @@ def get_common_params(self):
...
+                return int(output_lines[0])
+        else:
+            wc_params = ['wc', '-l']
+            wc_params.extend(input_filenames)
+            wc_output = command.execute_with_output(" ".join(wc_params))
+            # take the set of characters from the last line, which is the total number of lines
+            # for paired reads or the only line for unpaired reads
+            wc_lines = [line for line in wc_output.split("\n") if line != '']
+            wc_target_line = [line for line in wc_lines[-1].split(" ") if line != '']
+            total_line_count = ''.join(filter(lambda x: x.isdigit(), str(wc_target_line[0])))
```
Comment (MarkAZhang) [inline idseq_dag/steps/run_srst2.py]: Is there a chance that some of the characters won't be a number? Curious why you do the filter.

File: idseq_dag/steps/run_srst2.py [hunk 4]
```diff
@@ -69,6 +72,76 @@ def get_common_params(self):
...
+        """Gets the total number of reads in the sample by counting them directly from the
+            fastq or fasta files."""
+        input_filenames = self.input_files_local[0]
+        if is_zipped:
+            gunzip_params = ['gunzip', '-k']
+            gunzip_params.extend(input_filenames)
+            command.execute(" ".join(gunzip_params))
+            input_filenames = map(lambda name: name[:len(name)-3], input_filenames)
+        if is_fasta:
+            grep_params = ['grep', '-c', '"^>"'] # fastas start reads with "^>"
```
Comment (MarkAZhang) [inline idseq_dag/steps/run_srst2.py]: Is it not possible to do the `wc -l` approach for fasta files as well?

**Pull request closed without merging**